huge update read below

frankischilling opened the pull request:
## Summary - Huge UI, config, and file-workflow update for cupidfm

This pull request is a large quality-of-life and architecture update for cupidfm. It introduces a real configuration system, configurable keybindings, richer UI (notifications and banner handling), improved mime/emoji/nerdfont support, and a much more complete set of file operations, along with a long list of bug fixes and stability improvements.

* 129 commits
* 38 files changed
* 3 contributors

---

## Highlights

### 1. New configuration system (lib/cupidconf)

* Introduces a small configuration library in `lib/cupidconf.c` / `lib/cupidconf.h`.
* Supports:

  * Simple `key = value` pairs.
  * Comments starting with `#` and `;`.
  * Wildcard matching through `fnmatch` for both keys and values.
* API:

  * `cupidconf_load`, `cupidconf_get`, `cupidconf_get_list`.
  * `cupidconf_value_in_list` for checking whether a value matches any wildcard patterns under a key (useful for things like ignore lists).
* Centralizes config parsing so it can be reused by other Cupid tools in the future.

### 2. Configurable keybindings and defaults

* Adds `KeyBindings` struct and configuration loader in `src/config.c` / `src/config.h`.
* `load_default_keybindings` sets sane defaults for:

  * Navigation: arrows, Tab, exit.
  * File actions: edit, copy, paste, cut, delete, rename, new file, save, new directory.
  * Edit mode actions: movement, save (Ctrl+G), quit (Ctrl+Q), backspace.
* `load_config_file`:

  * Reads keybindings and `info_label_width` from a config file via cupidconf.
  * Parses human-readable values like:

    * `KEY_UP`, `KEY_DOWN`, `KEY_LEFT`, `KEY_RIGHT`
    * `F1`, `F2`, etc. and `KEY_F(n)`
    * `^C` style control keys
    * `Tab`, `Space`
    * `Shift+X` and a set of common shifted symbols
  * Accumulates helpful error messages in a buffer without crashing if parts of the config are invalid.
* Allows users to tailor keybindings and label width instead of recompiling.

### 3. New globals layer and banner/notification wiring

* Adds `src/globals.c` / `src/globals.h` to centralize shared state:

  * `BANNER_TEXT`, `BUILD_INFO`
  * `bannerwin`, `notifwin`
  * `last_scroll_time`, `last_notification_time`
  * `banner_mutex`, `banner_offset`
  * `KeyBindings g_kb`
  * Global flags such as `resized`, `is_editing`, `should_clear_notif`
* This de-spaghettis `main.c` and makes the UI state easier to reason about, especially around window resizing and edit mode.

### 4. UI and notification improvements

* Adds `src/ui.c` / `src/ui.h`:

  * `show_notification(WINDOW *win, const char *format, ...)` for inline messages in the notification bar.
  * `show_popup(const char *title, const char *fmt, ...)` for centered, blocking popup dialogs (used for confirmations and error messages).
* The rotating title banner:

  * Uses a shared `banner_offset` and a banner thread with mutex protection.
  * Now stays in sync with edit mode and terminal resize events (SIGWINCH).
  * Respects configuration so the banner can react to user config.

### 5. File editing, copy/cut/paste, and delete workflow

File operations have been significantly expanded:

* **Editing:**

  * In-terminal editor integrated with curses, with dedicated edit keybindings.
  * Fixes multiple crash and freeze issues:

    * Freeze on save in WSL.
    * Segfaults and crashes when editing empty files.
    * End-of-file handling issues.
* **Copy/Cut/Paste:**

  * Clipboard-style helpers in `src/utils.c`:

    * `copy_to_clipboard`
    * `paste_from_clipboard`
    * `cut_and_paste`
  * Paste can automatically add a numeric suffix when copying into a directory with existing files.
  * Fixes copy on folders and corner cases around cut/paste.
* **Delete:**

  * `delete_item` with proper confirmation via `confirm_delete`.
  * Delete shortcut wired into keybindings.
  * Delete confirmation added to prevent accidental removals.
* **Rename and creation:**

  * `rename_item` for inline renaming with feedback in the notification window.
  * `create_new_file` and `create_new_directory` for quick file/dir creation from within cupidfm.
  * Adds Shift support and a default shortcut for "new directory".
* **Directory navigation:**

  * Improved directory reload and navigation helpers, including better handling when moving right on non-directories.
  * Fixes multiple issues where files could break out of borders in the preview window and where cursor highlighting vanished at the bottom of the list.

### 6. Mime, emoji, nerdfont, and info display

* Uses `libmagic` to detect mime types, now wired through the updated build system.
* Adds emoji and Nerd Font icon support for file types:

  * Nerdfont tree preview for nicer directory trees.
  * Emoji fallback when nerdfonts are unavailable.
* Extends file info display:

  * Better mime-aware icons and labels.
  * `format_file_size` utility and safer usage guarded by proper checks.
  * Fixes bugs in `FileAttr_is_dir`, `display_file_info`, and size formatting.
* Introduces `mime_demo/` with sample files of various types (json, xml, text, c, js, py, shell, images, archive, audio, video) to verify mime detection and icon rendering.

### 7. Layout, resizing, and scroll fixes

* Multiple improvements around window layout and resizing:

  * Fixes resize issues that previously caused broken layouts.
  * Preview window and directory window now use more robust sizing logic.
  * Directory preview scrolling now works correctly when tabbed to the preview window.
  * Fixes banner scrolling and cursor highlight behavior when scrolling to the bottom of the directory list.
* Uses SIGWINCH handling and shared globals to ensure edit mode and normal mode both react correctly to terminal size changes.

### 8. CLI foundation

* Adds `src/cli.c` / `src/cli.h`:

  * `cli_readline` and `cli_println` helpers.
  * Lays groundwork for a future standalone CLI interface for cupidfm functionality without the curses UI.
* README updated with a "future for CLI" section so users know what is planned.

### 9. Refactors and internals

* Cleans up `Vector` and `VecStack`:

  * Switches from `nullptr` to `NULL` for C correctness.
  * Adds documentation and separates concerns for:

    * `VecStack_empty`, `VecStack_push`, `VecStack_pop`, `VecStack_peek`, `VecStack_bye`.
  * Adds capacity helpers like `Vector_min_cap` and `Vector_sane_cap`.
* Moves common macros and sizes (MAX, MIN, path lengths, label widths) to shared headers so they are defined in one place.
* Splits responsibilities across `globals.c`, `main.c`, `files.c`, `utils.c`, and `ui.c` for better maintainability.

### 10. Build and tooling changes

* `Makefile`:

  * Now links against `ncursesw` and `libmagic`.
  * Compiles `lib/cupidconf.c` directly into the binary.
  * Uses consistent include paths: `-Isrc -Ilib`.
* `dev.sh`:

  * Compiles with address sanitizer and undefined behavior sanitizer.
  * Explicitly links against `-lmagic` and `-lncursesw`.
  * Runs `./cupidfm 2> log.txt` after build for easier debugging.
* README:

  * Substantial expansion:

    * Updated build instructions for the new dependencies.
    * Configuration and keybinding documentation using cupidconf.
    * Screenshots (preview images added under `img/`) and usage examples.
    * License information added.
    * TODO list updated with current and future tasks.

---

## Bug fixes

This PR includes many targeted fixes that have accumulated over time:

* Fix freeze on save under WSL when editing files.
* Fix segfaults and crashes when:

  * Editing empty files.
  * Resizing terminal during edit mode.
  * Navigating past the last file.
* Fix files breaking out of the border in the preview window.
* Fix multiple resize issues and layout glitches.
* Fix usage of `format_file_size` without null checks.
* Fix `FileAttr_is_dir` and `display_file_info` logic.
* Fix directory preview scrolling when focus is on the preview window.
* Fix cursor highlighting disappearing after scrolling to the end of the directory listing.
* Improve WSL behavior for emojis and wide characters (see platform note below).

Thanks to @arthurbacci and @ferruccia1atwit for contributions around configuration safety and WSL fixes.

---

## Breaking or notable changes

* New runtime dependencies:

  * `ncursesw` for wide-character support.
  * `libmagic` for mime type detection.
* Configuration:

  * Keybindings and some layout options are now driven by user config files.
  * Users without a config file still get sane defaults; failures to load config are non-fatal and reported in an error buffer.
* Internal APIs and globals:

  * Some internal APIs and globals have moved; code that previously touched certain global UI variables directly should now use the new globals and helpers.
* Platform support:

  * Native Linux terminals remain the primary target for this release.
  * Due to ongoing problems with wide characters, emojis, and ghost characters in certain WSL terminal setups, WSL-specific behavior is considered **deprecated/unsupported for now**. Some WSL-oriented experiments were reverted; once a stable approach is found, WSL support can be revisited in a future PR.

---

## Checklist

Please update as you confirm:

* [x] Builds on modern Linux with `ncursesw` and `libmagic` installed.
* [ ] Basic navigation, edit, copy/cut/paste, rename, and delete tested.
* [x] Config file with custom keybindings and `info_label_width` tested.
* [x] README and screenshots match current behavior.

---

frankischilling committed:
basic file editing - freezing on save (WSL issue?)

frankischilling committed:
fix file saving with control+g

frankischilling committed:
Update README.md

frankischilling committed:
add notif window and change to use it

frankischilling committed:
updated readme with magiclib

frankischilling committed:
Update README.md

frankischilling committed:
mime type file editing, preview, etc

frankischilling committed:
Update dev.sh

frankischilling committed:
added resizing, fix right on non dir, fix on WSL

frankischilling committed:
fix seg fault

frankischilling committed:
Update README.md

frankischilling committed:
fix crash on editing empty file

frankischilling committed:
Update README.md

frankischilling committed:
Update README.md

frankischilling committed:
directory tree preview [wip]

frankischilling committed:
Update README.md

frankischilling committed:
update todo

frankischilling committed:
Update README.md

frankischilling committed:
better comments, some refactor

frankischilling committed:
Update README.md

frankischilling committed:
Update README.md

frankischilling committed:
Update README.md

frankischilling committed:
Update README.md

frankischilling committed:
rotating title banner

frankischilling committed:
Update README.md

frankischilling committed:
Update README.md

frankischilling committed:
Update main.c

frankischilling committed:
Update README.md

frankischilling committed:
Update main.c

frankischilling committed:
Update README.md

frankischilling committed:
fix resize

frankischilling committed:
Update README.md

frankischilling committed:
Update README.md

frankischilling committed:
fix resizing, preview win size, dir win size, 

also fixed and added a better banner for the rotating title

frankischilling committed:
Update README.md

frankischilling committed:
fix end of file, add makefile

frankischilling committed:
fix some files breaking out border in previw win

frankischilling committed:
Update README.md

frankischilling committed:
Update README.md

frankischilling committed:
Update README.md

frankischilling committed:
Update README.md

frankischilling committed:
fix issue #4

frankischilling committed:
try to work on banner not scrollin in edit mode

frankischilling committed:
fix issues

frankischilling committed:
fixes, and small additions

frankischilling committed:
nerdfonts support

frankischilling committed:
better performance, nerdfont tree preview

frankischilling committed:
better performance, mime icon types, mime demo

frankischilling committed:
emoji mime fallback

frankischilling committed:
Update Makefile

frankischilling committed:
Add files via upload

frankischilling committed:
add img

frankischilling committed:
Update README.md

frankischilling committed:
Update README.md

frankischilling committed:
fix emojis not being used in basic file info

frankischilling committed:
Update README.md

frankischilling committed:
Add files via upload

frankischilling committed:
Add files via upload

frankischilling committed:
Update README.md

frankischilling committed:
Update README.md

frankischilling committed:
Update README.md

frankischilling committed:
Update README.md

frankischilling committed:
Update README.md

frankischilling committed:
Update README.md

frankischilling committed:
fix a couple misc bugs

frankischilling committed:
Update README.md

frankischilling committed:
Update README.md

frankischilling committed:
Update README.md

frankischilling committed:
future for cli

frankischilling committed:
Update README.md

frankischilling committed:
Update README.md

frankischilling committed:
Update README.md

frankischilling committed:
Update README.md

frankischilling committed:
Update README.md

frankischilling committed:
Update README.md

frankischilling committed:
Update README.md

frankischilling committed:
Update README.md

frankischilling committed:
basic copy paste files

frankischilling committed:
fix copy on folder

frankischilling committed:
Update README.md

frankischilling committed:
Update README.md

frankischilling committed:
add cut and paste

frankischilling committed:
implement notif win time, delete shortcut

frankischilling committed:
basic dele

frankischilling committed:
fix cut and paste

frankischilling committed:
delete confirm

frankischilling committed:
Update README.md

frankischilling committed:
Update README.md

frankischilling committed:
rename + new file

frankischilling committed:
Delete preview2.png

frankischilling committed:
update readme for config

frankischilling committed:
config for keybindings

frankischilling committed:
config error handling, better parser, config #

frankischilling committed:
Update README.md

frankischilling committed:
banner reacts to config

frankischilling committed:
Update README.md

frankischilling committed:
Update README.md

frankischilling committed:
added shift support, create new dir

frankischilling committed:
Update README.md

frankischilling committed:
Update README.md

frankischilling committed:
Update README.md

frankischilling committed:
escape on prompt

frankischilling committed:
move stuff to globals

frankischilling committed:
format and comments, spring cleaning

frankischilling committed:
Update README.md

arthurbacci commented on src/cli.c:
```diff
@@ -1,28 +1,45 @@
 // File: cil.c
```
typo: should be cli.c

frankischilling replied:
fixed?

arthurbacci commented on src/config.c:
```diff
@@ -0,0 +1,328 @@
+    s = ltrim(s);  // Remove leading whitespace
+    rtrim(s);      // Remove trailing whitespace
+    return s;
+}
```
Could also be
```c
rtrim(s);
return ltrim(s);
```
, with one less line.

frankischilling replied:
fixed

arthurbacci commented on src/config.c:
```diff
@@ -0,0 +1,328 @@
+        char *eq = strchr(trimmed_line, '=');
+        if (!eq) {
+            // Malformed line without '='
+            if (error_count < buffer_size - 1) {
```
This if doesn't make sense for me

frankischilling replied:
The if (!eq) condition checks whether the = character exists in the configuration line. If eq is NULL (meaning no = was found), the line is considered malformed because it does not follow the expected key=value format. In that case, an error message is logged, the error count is incremented, and the program skips processing that line

arthurbacci commented on src/config.c:
```diff
@@ -0,0 +1,328 @@
+            continue;
+        }
+
+        // Map the parsed key code to the appropriate field in the KeyBindings structure
```
```c
#define CUPID_CFGCMP(x) else if (strcasecmp(name, #x) == 0) { kb->x = parsed };

if (false) {}
CUPID_CFGCMP(key_up) CUPID_CFGCMP(key_down) CUPID_CFGCMP(key_left)...
else {...

frankischilling replied:
added

arthurbacci commented on src/files.c:
```diff
@@ -34,11 +99,27 @@ const char *FileAttr_get_name(FileAttr fa) {
+    if (fa == NULL) {
+        return false; // Treat NULL as a non-directory
+    }
     return fa->is_dir;
```
`return fa != NULL && fa->is_dir;`

frankischilling replied:
fixed

arthurbacci commented on src/files.c:
Should be possible for the user to pick what he wants for this number

frankischilling replied:
user can now change it in config

arthurbacci commented on src/files.c:
```c
char fileSizeStr[20] = "Uncalculable";
if (dir_size != -2)
    format_file_size(fileSizeStr, dir_size)
mvwprintw(window, 2, 2, "%-*s %s", label_width, "📁 Directory Size:", fileSizeStr);
```

frankischilling replied:
fix

arthurbacci commented on src/files.c:
Awful repetition of this mvwprintw

frankischilling replied:
Fixed

arthurbacci commented on src/vecstack.c:
```diff
@@ -1,31 +1,67 @@
+    }
+
+    // Resize if necessary
+    if (stack->size == stack->capacity) {
```
using `>=` is a good practice

arthurbacci replied:
You really shouldn't be extending the capacity in the push function. I really dislike all changes made to this file as the code is bigger and more prone to errors. Is there a reason for not using the functions from vector.c like the old code?

frankischilling replied:
reverted

frankischilling committed:
Fix trim

frankischilling committed:
fix  FileAttr_is_dir

frankischilling committed:
Update README.md

frankischilling committed:
fix display_file_info

frankischilling committed:
ability to config label_width (wont change much)

config parser works with info_label_width or label_width

frankischilling committed:
fixes

arthurbacci committed:
fix lack of if before usage of format_file_size

frankischilling committed:
revert to old vecstack code

ferruccia1atwit committed:
fix on ubuntu wsl, fix default config typ

frankischilling committed:
Merge pull request #7 from ferruccia1atwit/file-editing

fix on ubuntu wsl, fix default config typ

frankischilling committed:
macro for config, paste will add number if copy

frankischilling committed:
revert mess up

frankischilling committed:
move build instructs up

frankischilling committed:
Switch to ncursesw for proper Unicode support

frankischilling committed:
revert until fix ghost char issue when using (maybe wsl)

frankischilling committed:
Fixes emoji support for WSL, causes other

An issue as been made for this bug

frankischilling committed:
use cupidconf

frankischilling committed:
Update README.md

frankischilling committed:
Update TODO list in README.md

frankischilling committed:
Fix directory preview scrolling when tabbed to preview window

Added scroll support for directory tree previews. Users can now
scroll through directory contents when the preview window is active
using arrow keys, just like file previews.

- Added directory tree line counting functions
- Modified show_directory_tree() to support scroll offsets
- Updated preview window drawing and scrolling logic to handle both files and directories

Fixes: Directory preview not scrolling when tabbed to preview window

frankischilling committed:
Update README with completed task and license info

frankischilling committed:
Fix cursor highlighting disappearing when scrolling to bottom of directory list

Fixed an inconsistency in the num_lines calculation for the directory window
that caused the cursor highlighting to become invisible when scrolling past
the last visible item in a long directory list.

The issue was caused by num_lines being set to LINES - 5 in some places
while the actual window height was LINES - 6. This mismatch caused the
scrolling logic to miscalculate the visible area, allowing the cursor to
be positioned outside the visible window bounds.

Changes:
- Updated num_lines calculation from LINES - 5 to LINES - 6 in:
  * navigate_left() function
  * navigate_right() function  
  * main() initial state setup

This ensures the cursor position calculations are consistent with the actual
window dimensions (banner: 3 lines, notification: 1 line, main window borders:
2 lines), keeping the cursor highlighting visible at all times.

frankischilling committed:
Update README with completed tasks and formatting fixes

Updated README to reflect completed tasks and fixed formatting issues.

frankischilling committed:
Fix edit mode issues: banner synchronization, SIGWINCH handling, and text buffer improvements

Fixed multiple issues related to edit mode functionality, banner marquee synchronization,
and text buffer scrolling behavior.

Banner Marquee Synchronization:
- Fixed banner marquee timing desynchronization when entering/exiting edit mode or prompts
- Replaced static banner_offset variables with a shared global banner_offset
- All functions (main loop, edit mode, prompts) now use the same banner_offset
- Banner continues from the same position when entering prompts or edit mode
- Prevents banner from jumping to different positions after exiting prompts/edit mode

SIGWINCH Handling in Edit Mode:
- Fixed window resize signal handling breaking while in edit mode
- Removed is_editing check in handle_winch() to allow resize signals during edit mode
- Added proper resize handling in edit mode loop with window recreation
- Banner and notification windows are now properly resized during edit mode
- Editor window is recreated with correct dimensions on resize

Text Editor Cursor and Display:
- Fixed cursor appearing at bottom of text buffer and near marquee
- Changed from terminal cursor (curs_set(1)) to visual highlighting only (curs_set(0))
- Replaced wmove() + waddch() with mvwaddch() to prevent cursor positioning issues
- Cursor now only appears as highlighted character at correct position

Text Buffer Scrolling:
- Fixed text buffer not using full available horizontal space
- Added logic to minimize horizontal scroll when window gets wider
- Fixed horizontal scroll reset when window width becomes too small
- Added safeguards for very narrow windows to prevent scroll calculation errors
- Text buffer now properly adjusts to use maximum available space

Window Redraw on Exit:
- Added full window redraw when exiting edit mode
- Sets resized flag to trigger redraw_all_windows() in main loop
- Ensures all windows are properly redrawn with borders after edit mode
- Prevents leftover text and missing borders after exiting editor

Files modified:
- src/files.c: Edit mode improvements, cursor handling, scroll management
- src/main.c: Banner offset synchronization, window redraw on exit
- src/utils.c: Banner offset synchronization in prompt functions
- src/globals.h: Added global banner_offset declaration
- src/globals.c: Added global banner_offset definition

frankischilling commented:
Too much to write to explain what was added, a lot.

frankischilling committed:
Refactor MIME type display handling

Refactor MIME type display logic to use a single variable for output, simplifying code and ensuring proper truncation.

frankischilling committed:
Merge branch 'main' into file-editing

frankischilling committed:
Revise README with updates and issue fixes

Updated README.md to reflect changes in features, fixed issues, and removed outdated usage examples.
